### PR TITLE
Add `vhost-user-sound` crate

### PR DESCRIPTION
**Pull request by epilys**
### Summary of the PR

This PR adds a new crate, `vhost-user-sound`, that provides a `vhost-user` backend daemon for the VIRTIO Sound device as specified in the VIRTIO Spec v.1.2 [^0]

Supported features are:

- One input stream, one output stream (hard-coded).
- Playback with ALSA backend (on host machine).
  Buffer/period settings need some tuning, the first period seems to come up short; I'm not familiar with ALSA programming at all.
- Playback with a "null" backend that just consumes audio without actual playback.

What remains to be done:

- There are various places where micro-refactors make sense
- Audio capture
- Add more audio backends.
- Add logic for setting up the default sound device stream information on startup from the audio backend, instead of hardcoding one input stream and one output stream.

[^0]:  [Virtual I/O Device (VIRTIO) Version 1.2 Committee Specification Draft 01 09 May 2022 ](https://docs.oasis-open.org/virtio/virtio/v1.2/csd01/virtio-v1.2-csd01.html#x1-52900014)

### Usage / Testing functionality

This PR was tested with QEMU, with a proposed patchset that enables generic `vhost-user` devices in virtualized guests [^1], contributed by @stsquad.

[^1]: https://patchew.org/QEMU/20230418162140.373219-1-alex.bennee@linaro.org/

The following command line invocations were used:

#### Run the `vhost-device` daemon

```shell
cargo run -- --socket /tmp/snd.sock --backend alsa
```

#### Run a guest in QEMU

1. Apply the patchset [^1].
3. Apply these patches to create a `vhost-user-snd` and `vhost-user-snd-pci` device [^2].

   - [`[PATCH] Add virtio-sound and virtio-sound-pci devices`](https://github.com/epilys/qemu-virtio-snd/commit/6fbca9fe0396c473bb16726f3d10bd4b5483447e.patch)
   - [`[PATCH] Add vhost-user-snd and vhost-user-snd-pci devices`](https://github.com/epilys/qemu-virtio-snd/commit/54ae1cdd15fef2d88e9e387a175f099a38c636f4.patch)

   [^2]: <https://github.com/epilys/qemu-virtio-snd/commits/54ae1cdd15fef2d88e9e387a175f099a38c636f4>

2. Build for desired target architecture, e.g. `x86_64`.
3. Get/build kernel configured with `CONFIG_SND_VIRTIO`.
4. Launch QEMU with

   ```shell
   $ export VHOST_USER_SOCK=/tmp/snd.sock
   $ /path/to/qemu-system-x86_64  \
       [... snip ...] \
     -chardev socket,id=vsnd,path="$VHOST_USER_SOCK" \
     -device vhost-user-snd-pci,chardev=vsnd,id=snd \
     [... snip ...]
   ```
5. Use `speaker-test -l 1` to play some pink noise for one loop (`-l N`) or `speaker-test -t wav -w /path/to/file.wav -l 1`.

### Requirements

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: crates/sound/src/device.rs [hunk 1]
```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
...
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0).into());
+            }
+
+            let request = desc_chain
+                .memory()
+                .read_obj::<VirtioSoundHeader>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            // Keep track of bytes that will be written in the VQ.
+            let mut len = 0;
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs]: What about renaming in `used_len`?

File: crates/sound/src/device.rs [hunk 2]
```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
...
+
+            // Reply header descriptor.
+            let desc_hdr = descriptors[1];
+            if !desc_hdr.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(1).into());
+            }
+            len += desc_hdr.len();
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs]: I would move this when we really write the descriptor, just to be more readable.

File: crates/sound/src/device.rs [hunk 3]
```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
...
+                        .read_obj::<VirtioSoundQueryInfo>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+
+                    let start_id = u32::from(request.start_id) as usize;
+                    let count = u32::from(request.count) as usize;
+                    let streams = self.streams.read().unwrap();
+                    if streams.len() <= start_id || streams.len() <= start_id + count {
+                        resp.code = VIRTIO_SND_S_BAD_MSG.into();
+                    } else {
+                        let desc_response = descriptors[2];
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs:204]: Are we sure that we always have this layout?

I mean: descriptor[1] header and descriptor[2] payload.

Or is it an implementation detail of the linux driver?

I ask because in virtio-vsock we had a similar problem: the specification doesn't define these details and the driver initially used 2 separate descriptors for header and payload, then it started to use only one descriptor and vhost-user-vsock didn't work.

That's fine for now, but I would put a comment if that's the case.

File: crates/sound/src/device.rs [hunk 4]
```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
...
+
+        // Reply to some requests right away, and defer others to the audio backend (for
+        // example PcmRelease needs to complete all I/O before replying.
+        //
+        // Mark `any` as true if we need to reply to any request right away so we can
+        // signal the queue as used.
+        let mut any = false;
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs:144]: Should we check that the length of `descriptors` is what we expect?

File: crates/sound/src/device.rs [hunk 5]
```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
...
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            if vring.add_used(desc_chain.head_index(), len).is_err() {
+                log::error!("Couldn't return used descriptors to the ring");
+            }
+            any |= true;
+        }
+
+        // Send notification if any request was processed
+        if any && vring.signal_used_queue().is_err() {
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs:366]: In which cases can happen that we get here and `any` is false?

File: crates/sound/src/device.rs [hunk 6]
```diff
@@ -171,7 +456,7 @@ impl VhostUserBackend<VringRwLock, ()> for VhostUserSoundBackend {
     }
 
     fn max_queue_size(&self) -> usize {
-        256
+        64
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs:589]: Why this change?

**Comment by epilys** [comment]
@stefano-garzarella This is the PR as it was rebased against current upstream so it shows unrelated commits (I have to re-rebase it against this repo)

Should I make a new PR with only the alsa backend? I see there is code and open PRs handling TX and streams already so the alsa backend is the biggest difference.

File: crates/sound/src/lib.rs [hunk 7]
```diff
@@ -224,16 +223,39 @@ impl SoundConfig {
...
+        if self
+            .vring
+            .add_used(self.desc_chain.head_index(), resp.as_slice().len() as u32)
+            .is_err()
+        {
+            log::error!("Couldn't add used");
+        }
+        if self.vring.signal_used_queue().is_err() {
+            log::error!("Couldn't signal used queue");
+        }
```
Comment (stefano-garzarella) [inline crates/sound/src/lib.rs:258]: The main drawback of this approach (which I don't dislike) is that we don't return an error via `process_*()`, right?

**Comment by stefano-garzarella** [comment]
> @stefano-garzarella This is the PR as it was rebased against current upstream so it shows unrelated commits (I have to re-rebase it against this repo)

Yep, would be great if you can rebase on top of `virtio-sound` branch.
Anyway, I also did a quick review and left some comments.

> 
> Should I make a new PR with only the alsa backend? I see there is code and open PRs handling TX and streams already so the alsa backend is the biggest difference.

@MatiasVara @dorindabassey WDYT?

File: crates/sound/src/device.rs (hunk 1 above)
Comment (epilys) [reply crates/sound/src/device.rs]: +1 will do.

File: crates/sound/src/device.rs (hunk 3 above)
Comment (epilys) [reply crates/sound/src/device.rs:204]: I think so, at least from my understanding of the spec wording (I haven't worked with virtqueues directly before, caveat emptor)

```
A generic control message consists of a request part and a response part.

A request part has, or consists of, a common header containing the following device-readable field:

code
    specifies a device request type (VIRTIO_SND_R_*).

A response part has, or consists of, a common header containing the following device-writable field:

code
    indicates a device request status (VIRTIO_SND_S_*).

The status field can take one of the following values:

    VIRTIO_SND_S_OK - success.
    VIRTIO_SND_S_BAD_MSG - a control message is malformed or contains invalid parameters.
    VIRTIO_SND_S_NOT_SUPP - requested operation or parameters are not supported.
    VIRTIO_SND_S_IO_ERR - an I/O error occurred.

The request part may be followed by an additional device-readable payload, and the response part may be followed by an additional device-writable payload. 
```

So in this case the generic ctrl header is followed by the additional device-readable payload, and then the response comes after.

File: crates/sound/src/device.rs (hunk 4 above)
Comment (epilys) [reply crates/sound/src/device.rs:144]: I wasn't exactly sure on this because the spec says that the header is followed by the buffer which may be of any size * therefore we don't know how many descriptors, except for the last one which would be the response/status. So if we read we IO request header, we should expect the last descriptor to be the response and any descriptor before that a buffer:

```
An I/O message consists of the header part, followed by the buffer, and then the status part. 
```

* EDIT: any size less than period_bytes but that's a SHOULD not a must:

> The driver SHOULD populate the tx queue with period_bytes sized buffers. The only exception is the end of the stream.

File: crates/sound/src/device.rs (hunk 5 above)
Comment (epilys) [reply crates/sound/src/device.rs:366]: If the request buffer is empty, there should be an empty descriptor or lack of descriptors(?), which the spec doesn't explicitly state (Again, I think, I am not exactly sure) the relevant part is this:

> An I/O message consists of the header part, followed by the buffer, and then the status part. 

I'd expect there would be an empty descriptor instead of none but I couldn't find any such guarantee in the code, or the spec

File: crates/sound/src/device.rs (hunk 6 above)
Comment (epilys) [reply crates/sound/src/device.rs:589]: Oh I wanted input on this actually. Doesn't this specify the max number of descriptors? (`The number of descriptors in the table is defined by the queue size for this virtqueue: this is the maximum possible descriptor chain length.`) I don't know why, but `crosvm` set this value.

File: crates/sound/src/device.rs (hunk 3 above)
Comment (stefano-garzarella) [reply crates/sound/src/device.rs:204]: Okay, so even in this case it's an implementation detail of the driver. Nothing prevents it from using a single writable descriptor to contain both the header and the response payload, right?

File: crates/sound/src/lib.rs (hunk 7 above)
Comment (epilys) [reply crates/sound/src/lib.rs:258]: Indeed, I took this approach from crates/gpio/src/vhu_gpio.rs and I'm not sure what to do here. I'll have to check what the vhost-user backend trait does with the errors returned by the handling code

File: crates/sound/src/device.rs (hunk 5 above)
Comment (epilys) [reply crates/sound/src/device.rs:366]: this might be clarified somewhere else that specifies the mapping of payload to virt queue descriptors.

File: crates/sound/src/device.rs (hunk 2 above)
Comment (epilys) [reply crates/sound/src/device.rs]: +1 will do.

File: crates/sound/src/device.rs (hunk 3 above)
Comment (Ablu) [reply crates/sound/src/device.rs:204]: Almost all rust-vmm daemons get this wrong in one way or another... We would really need something like https://github.com/rust-vmm/vm-virtio/pull/33 to prevent daemons from needing to worry about this stuff...

File: crates/sound/src/device.rs (hunk 4 above)
Comment (stefano-garzarella) [reply crates/sound/src/device.rs:144]: Yep, we should have a generic code that can handle any number of descriptor. But IIUC, none of the rust-vmm/vhost-user devices do that. So, I 'd put comments to specify these things, in this way in the future if we have problems, we know why.

I mean for now is fine to fix the number of descriptor, based on the current implementation of Linux's driver.

In the future I think we should do something like this (not yet implemented for vsock_packet): https://github.com/rust-vmm/vm-virtio/issues/216

Comment (stefano-garzarella) [reply crates/sound/src/device.rs:144]: In this case I meant that we access `descriptor[1]` and `descriptor[2]` for example. Should we check the length of the vector first to avoid panics at runtime?

File: crates/sound/src/device.rs (hunk 3 above)
Comment (stefano-garzarella) [reply crates/sound/src/device.rs:204]: Yep, I came to the same conclusion.

So that's okay in my opinion for now, but we should put comments where we make this assumption to say that's how the driver works in Linux 6.4.

WDYT?

File: crates/sound/src/device.rs (hunk 6 above)
Comment (stefano-garzarella) [reply crates/sound/src/device.rs:589]: Right, I don't know if we have to make it parametric, but maybe the driver in Linux always uses 64 and they put it that way. I might have picked it up from vhost-user-vsock. We need to understand it better, however I don't think it's a big problem.

File: crates/sound/src/device.rs (hunk 3 above)
Comment (Ablu) [reply crates/sound/src/device.rs:204]: > So that's okay in my opinion for now, but we should put comments where we make this assumption to say that's how the driver works in Linux 6.4.

Sounds fine to me at least. Everyone else seems to do the same, just without the comment :).

Comment (epilys) [reply crates/sound/src/device.rs:204]: > Nothing prevents it from using a single writable descriptor to contain both the header and the response payload, right?

It does say each "element" in the virtqueue is placed on a separate descriptor but could definitely be more specific:

```
> 2.7.13.1 Placing Buffers Into The Descriptor Table

A buffer consists of zero or more device-readable physically-contiguous elements followed by zero or more physically-contiguous device-writable elements (each has at least one element). This algorithm maps it into the descriptor table to form a descriptor chain:

for each buffer element, b:

    Get the next free descriptor table entry, d
    Set d.addr to the physical address of the start of b
    Set d.len to the length of b.
    If b is device-writable, set d.flags to VIRTQ_DESC_F_WRITE, otherwise 0.
    If there is a buffer element after this:
        Set d.next to the index of the next free descriptor element.
        Set the VIRTQ_DESC_F_NEXT bit in d.flags.

In practice, d.next is usually used to chain free descriptors, and a separate count kept to check there are enough free descriptors before beginning the mappings.
```

Comment (epilys) [reply crates/sound/src/device.rs:204]: I think this means it's not implementation specific

File: crates/sound/src/device.rs (hunk 4 above)
Comment (epilys) [reply crates/sound/src/device.rs:144]: that's how I had it at first, but then I found out the kernel times out on most control messages (I mentioned it in the last sync meeting) and the release response must wait for IO to be completed first. So some times we need to let the audio backend drop the control message and signal_used the queue. It could definitely be handled per message instead of after the `match {}`. I don't know if there's a meaningful difference between the two

File: crates/sound/src/device.rs (hunk 3 above)
Comment (Ablu) [reply crates/sound/src/device.rs:204]: I digged this out of my IRC history:
```
Ablu: The "Supplying Buffers to The Device" subsections for both splitted and packaged virtqueues seems to suggest that a single buffer can be split across multiple descriptors... 
vireshk: Do you have a reference to something that explains the 1 buffer == 1 descriptor? I guess naturally, most of the time that will be the case, just trying to read what the spec says literally...

Ablu: I think I found it... https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-280004 says that it can be split up arbitrarily. So I think the GPIO device, should (in theory) also be able to handle >2 descriptors I think? Though then... https://github.com/rust-vmm/vm-virtio/blob/main/crates/devices/virtio-blk/src/request.rs#L236-L244 also looks "wrong" since nothing specifies that the status has to be in it's own descriptor? So chances are that I am still missing something...

Ablu: I asked on rust-vmm slack about the virtio-blk code... 

Ablu: https://github.com/rust-vmm/vm-virtio/issues/216 is the answer :). So I am tempted to claim that the GPIO code asserting len == 2 is "wrong".
```
Unfortunately the Slack thread about this got lost since it is older than 90 days. But maybe this gives some context :).

Comment (stefano-garzarella) [reply crates/sound/src/device.rs:204]: @epilys header and payload could be on a single buffer, though, couldn't they?

The important thing is not to mix writable and readable buffers, but then for virtqueue it can contain anything as long as it's contiguous memory. So nothing should prevent the driver from using a single contiguous memory area and expecting the device to write both the header and the response payload there.

Comment (epilys) [reply crates/sound/src/device.rs:204]: are they considered the same element? (response header + payload)

Comment (Ablu) [reply crates/sound/src/device.rs:204]: Yes, I think what you quoted above talks about an "element" in the virtqueue. So the element there is a descriptor chain. The chain then can have any form (as long as r/w constraints are satisfied) and consist of multiple descriptors.

Comment (stefano-garzarella) [reply crates/sound/src/device.rs:204]: an element for the virtqueue are just contiguous bytes, then as each device uses them, they are a higher level of abstraction

File: crates/sound/src/device.rs (hunk 4 above)
Comment (stefano-garzarella) [reply crates/sound/src/device.rs:144]: @epilys sorry, I'm bit lost here :-( Is this related to the length of the vector?

File: crates/sound/src/device.rs (hunk 5 above)
Comment (MatiasVara) [reply crates/sound/src/device.rs:366]: If I understand correctly, any is `false` if the code does not go into the for-loop (https://github.com/epilys/vhost-device/blob/95ca8a85f12bc52946e022c9d94cf89d75d40666/crates/sound/src/device.rs#L143). However, I wonder if the codes does not exit before at https://github.com/epilys/vhost-device/blob/95ca8a85f12bc52946e022c9d94cf89d75d40666/crates/sound/src/device.rs#L132.

File: crates/sound/src/device.rs (hunk 3 above)
Comment (epilys) [reply crates/sound/src/device.rs:204]: I see, thanks!

File: crates/sound/src/device.rs (hunk 4 above)
Comment (epilys) [reply crates/sound/src/device.rs:144]: @stefano-garzarella My bad I mixed up what queue handler we were talking about.

So descriptors.len() should be at least two, the request and the response so definitely we should bail out early if the length is less than 2.

Comment (stefano-garzarella) [reply crates/sound/src/device.rs:144]: > @stefano-garzarella My bad I mixed up what queue handler we were talking about.

Got it. IIUC you were referring to `IOMessage::drop`, right?
So this could be another drawback, maybe for now it is better to do it the traditional way and move that code directly into `process_*()`. The idea isn't bad though, except that we actually might introduce hidden problems since we don't have precise control over the drop.

> So descriptors.len() should be at least two, the request and the response so definitely we should bail out early if the length is less than 2.

Yep, I agree.

Comment (epilys) [reply crates/sound/src/device.rs:144]: IO messages are async, so we can't move the TX/RX response/signal logic back to process_tx because the thread running it would have to block waiting for the audio backend to process the IO. The way I structured IOMessage is basically reference counted garbage collection, every `Buffer` holds a reference to `IOMessage` via `Arc`, and when all `Buffer`s are dropped so is `IOMessage`, which cleans up the request.

Comment (stefano-garzarella) [reply crates/sound/src/device.rs:144]: @epilys Sorry I thought we were talking about the control ones. We are overlapping too many things ;-) anyway now I understand, yes it makes sense!

Comment (MatiasVara) [reply crates/sound/src/device.rs:144]: @epilys What do you mean by `IO messages are async`? Are you referring to the specification that states that buffers should be marked used immediately after consumption?
```The device MUST NOT complete the I/O request until the buffer is totally consumed.``

Comment (epilys) [reply crates/sound/src/device.rs:144]: > the completion of such an I/O request can be considered an elapsed period notification. 

Consuming a TX I/O message at once can make the guest kernel think the period has elapsed therefore it should write more bytes into the sound card.

Comment (MatiasVara) [reply crates/sound/src/device.rs:144]: Oh, I see, thanks for point it out. Does this mean the device is returning used buffers later than expected by the kernel?

Comment (epilys) [reply crates/sound/src/device.rs:144]: No, the kernel writes data to a soundcard as fast as it can consume them (For buffering). The spec says that completion of the request means the hardware is ready for more buffering, analogous to hardware issuing a new interrupt (that's what a period refers to)

Comment (MatiasVara) [reply crates/sound/src/device.rs:144]: Thanks. In my comment, I meant the kernel guest. I was wondering how much later we are returning used buffers compared to the period. Considering that you mentioned the guest kernel assumes the period has expired when it returns a consumed buffer in the tx, I was wondering what the upper bound would be when that response is delayed.

Comment (epilys) [reply crates/sound/src/device.rs:144]: I was talking about the guest kernel as well

**Comment by epilys** [comment]
Didn't rebase yet against `virtio-sound` because it has an error: the AudioBackend trait uses a struct defined in the pipewire backend which is behind a cfg compile-time flag.

**Comment by dorindabassey** [comment]
> Didn't rebase yet against `virtio-sound` because it has an error: the AudioBackend trait uses a struct defined in the pipewire backend which is behind a cfg compile-time flag.

is this the struct you are referring to? https://github.com/virtio-sound/vhost-device/blob/3a0f49d27977188e1c3aa025cdd5185f04e71b22/crates/sound/src/audio_backends/pw_backend.rs#L14 
I think the next PR fixes that, so once we merge it, I think you should be able to rebase on top of it.

**Comment by dorindabassey** [comment]
I wonder if it make sense to you to split the PR into maybe 3 or 4, it depends - (documentation, "changes for ctrl queue, tx and rx queue that you think is nice to have", and the alsa-Backend.) IMHO that would simplify the review and merge process.

**Comment by epilys** [comment]
@dorindabassey you can view each commit separately, there are standalone commits for CTRL queue, TX queue and the alsa backend. Click on the files changed tab and select "view changes for X/Y/Z commit"

**Comment by dorindabassey** [comment]
> Didn't rebase yet against `virtio-sound` because it has an error: the AudioBackend trait uses a struct defined in the pipewire backend which is behind a cfg compile-time flag.

Hi Manos, Everything is okay now, I think you can rebase your PR on top of our current work. Just a small note - please take care with the rebasing so that the already existing changes do not get lost. thank you :)
